modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,295 @@
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param ILO_URL_Metric $new_url_metric New URL metric.
+	 */
+	public function add( ILO_URL_Metric $new_url_metric ) {
+		$url_metrics = $this->flatten();
+		array_unshift( $url_metrics, $new_url_metric );
+
+		$grouped_url_metrics = $this->group_url_metrics_by_breakpoint( $url_metrics );
+
+		// Make sure there is at most $sample_size number of URL metrics for each breakpoint.
+		$grouped_url_metrics = array_map(
+			function ( $breakpoint_url_metrics ) {
+				if ( count( $breakpoint_url_metrics ) > $this->sample_size ) {
+
+					// Sort URL metrics in descending order by timestamp.
+					usort(
+						$breakpoint_url_metrics,
+						static function ( ILO_URL_Metric $a, ILO_URL_Metric $b ): int {
+							return $b->get_timestamp() <=> $a->get_timestamp();
+						}
+					);
+
+					// Only keep the sample size of the newest URL metrics.
+					$breakpoint_url_metrics = array_slice( $breakpoint_url_metrics, 0, $this->sample_size );
+				}
+				return $breakpoint_url_metrics;
+			},
+			$grouped_url_metrics
+		);
+
+		$this->groups = $grouped_url_metrics;
+	}
+
+	/**
+	 * Groups URL metrics by breakpoint.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics URL metrics.
+	 * @return array<int, ILO_URL_Metric[]> URL metrics grouped by breakpoint. The array keys are the minimum widths for a viewport to lie within
+	 *                                      the breakpoint. The returned array is always one larger than the provided array of breakpoints, since
+	 *                                      the breakpoints reflect the max inclusive boundaries whereas the return value is the groups of page
+	 *                                      metrics with viewports on either side of the breakpoint boundaries.
+	 */
+	private function group_url_metrics_by_breakpoint( array $url_metrics ): array {
+
+		// Convert breakpoint max widths into viewport minimum widths.
+		$minimum_viewport_widths = array_map(
+			static function ( $breakpoint ) {
+				return $breakpoint + 1;
+			},
+			$this->breakpoints
+		);
+
+		$grouped = array_fill_keys( array_merge( array( 0 ), $minimum_viewport_widths ), array() );
+
+		foreach ( $url_metrics as $url_metric ) {
+			$current_minimum_viewport = 0;
+			foreach ( $minimum_viewport_widths as $viewport_minimum_width ) {
+				if ( $url_metric->get_viewport()['width'] > $viewport_minimum_width ) {
+					$current_minimum_viewport = $viewport_minimum_width;
+				} else {
+					break;
+				}
+			}
+
+			$grouped[ $current_minimum_viewport ][] = $url_metric;
+		}
+		return $grouped;
+	}
+
+	/**
+	 * Gets needed minimum viewport widths.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param float $current_time Current time, defaults to `microtime(true)`.
+	 * @return array<int, array{int, bool}> Array of tuples mapping minimum viewport width to whether URL metric(s) are needed.
```

Comment:
I think a simpler API, which would also be better in line with the method name, would be to return a simple indexed array of the minimum viewport widths.

I'm not sure there is a need for this to return all minimum viewport widths with a boolean indicator.

Reply:
It took me longer than I wished to prove this out but I believe I've finally done so. In db8ae4a6a0802b2fc964fc4f21bfa3be6200e6fd I added a test which:

1. Establishes a single breakpoint for which there are two viewport groups.
2. Fully populates the group for the wider viewports and leaves the group empty for narrower viewports.
3. Attempts to over-populate the wider viewport group with yet another URL metric.

When switching to only return the simple indexed array of the minimum viewport widths needing URL metrics (as I did in 794061d10183286ec7b9041f5a2e54cc8a4d37af), the REST API endpoint [fails](https://github.com/WordPress/performance/actions/runs/8103921850/job/22149500014) to properly block attempts to overpopulate the wider viewport group. The reason is that the narrower viewport group is encountered first when iterating which then sets `$last_was_needed` to `true`:

https://github.com/WordPress/performance/blob/794061d10183286ec7b9041f5a2e54cc8a4d37af/modules/images/image-loading-optimization/storage/rest-api.php#L128-L137

It never even considers the wider viewport group since it isn't returned by `get_needed_minimum_viewport_widths()`. Contrast this with the previous working code:

https://github.com/WordPress/performance/blob/ceeab6ec00024a1bf87e57da164e6284b67fd1ca/modules/images/image-loading-optimization/storage/rest-api.php#L128-L137

Note that in this case it iterates over _all_ viewport widths ensuring that the `$last_was_needed` can be switched back to `false` to block a submission.

Reply:
Did you see my comment https://github.com/WordPress/performance/pull/1010#discussion_r1506883862? Wouldn't something like that address the problem in a more elegant way? We don't need to get a list that includes all viewports to find out whether the viewport we already know (as specified in the request) is full.

Looking at my previous suggestion again, I think I missed one thing though, the minimum viewport width is of course not necessarily the exact viewport width provided. So we need a method that, for a given viewport width, returns the appropriate breakpoint / minimum viewport width that identifies the relevant group. Still, I think we could simplify this via a method in the class that checks for a given viewport width whether its information is needed.

For even more separation of concerns, this could work as follows:
```php
/*
 * Get the correct breakpoint / minimum viewport width that the given viewport width belongs to.
 * For example, for breakpoints 0, 600, and 1200:
 * - 599 would return 0
 * - 1199 would return 600
 * - 1200 would return 1200
 * - 8000 would return 1200
 */
$minimum_viewport_width = $grouped_url_metrics->get_minimum_viewport_width( $viewport_width );

/*
 * If the group for that breakpoint / minimum viewport width is already complete/full
 * (i.e. as large as the sample size), fail with an error.
 */
if ( $grouped_url_metrics->is_group_complete( $minimum_viewport_width ) ) {
	return new WP_Error( '...' );
}
```

This would abstract away the logic so we wouldn't need to have a complicated `foreach` in here.

Reply:
By the way, I wonder whether in terms of naming we should use "breakpoint" instead of "minimum viewport width". They are not exactly the same technically, but "breakpoint" is just a lot more concise for all those method and variable names. Plus no confusion with a viewport (i.e. a concrete viewport that isn't just the "minimum viewport" for a group).

Reply:
Yes, the logic can be encapsulated into a method to avoid having to iterate over the values in the REST API handler. I wanted to first sanity check why `get_needed_minimum_viewport_widths()` can't just return an array of integers for the viewport widths that actually need URL metrics.

Regarding the naming to use breakpoint, it can get a bit tricky because the breakpoints are the division between the groups, whereas the minimum viewport widths correspond to the number of groups themselves. The breakpoint count is always one less than the number of minimum viewport widths. 

Let me think about how this method might be named a bit more.

Reply:
I've introduced an `is_group_filled` method in b7d0819f.

Reply:
Regarding "breakpoint" vs "minimum viewport width", the difference you've explained makes sense. I personally still feel using the term "breakpoint" is fine and is just more approachable from a code perspective. While `0` isn't really a "breakpoint", we're talking about "grouping by breakpoint" in a few places, so every group represents a specific breakpoint. So from that perspective I'd also be inclined to go with the much shorter "breakpoint" terminology.

Reply:
Not sure I understand. Are you saying so far breakpoints are the "maximum widths" rather than the "minimum widths of the next larger group"?

Alternatively, how about "breakpoint groups" as terminology? Or "group IDs"? I just want to avoid the excessively long "minimum viewport widths".

Separately, WDYT about https://github.com/WordPress/performance/pull/1010#discussion_r1508340377? Looking at https://github.com/WordPress/performance/pull/1010/commits/b7d0819f3d2fe9d5fe9c0056caa8f0c1766e13ca#diff-4369a8ab8272ee7760fdc9ce4442c633c520114a2d593c859b7807e7ddb806dcR163, I still don't understand why we have to call that method here. IMO it would be more intuitive to have one method to get the correct "breakpoint group ID" or "minimum viewport width" for the group that the given viewport width falls into, and another method to check whether that group is filled.

Reply:
Breakpoints are the maximum widths before entering the next group. Given 3 breakpoints 480, 600, and 782 then there would be 4 groups for the following viewport ranges:

* 0-480
* 481-600
* 601-782
* 782-∞

So there is always one more group than there are breakpoints. I organized around "minimum viewport widths" since it makes sense to anchor off of the zero value, rather than to work from the other end at infinity (or `PHP_INT_MAX`). I don't have a problem with "minimum viewport widths" as a name, as it is much more explicit at what it is talking about. Another name like "group ID" or "breakpoint group" is more general and you'd have to remind yourself what it is referring to.

https://github.com/WordPress/performance/blob/3523e98b62ce2627502f1b75d28c5de1cee48023/modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php#L132-L140

> Separately, WDYT about [#1010 (comment)](https://github.com/WordPress/performance/pull/1010#discussion_r1508340377)? Looking at [b7d0819#diff-4369a8ab8272ee7760fdc9ce4442c633c520114a2d593c859b7807e7ddb806dcR163](https://github.com/WordPress/performance/commit/b7d0819f3d2fe9d5fe9c0056caa8f0c1766e13ca#diff-4369a8ab8272ee7760fdc9ce4442c633c520114a2d593c859b7807e7ddb806dcR163), I still don't understand why we have to call that method here. IMO it would be more intuitive to have one method to get the correct "breakpoint group ID" or "minimum viewport width" for the group that the given viewport width falls into, and another method to check whether that group is filled.

It could be refactored to access the data in a different way specifically for the purposes of how the data is needed by the REST API here, to specifically look up one single viewport width to see if its group is filled or not. However, the data is needed as well in `ilo_optimize_template_output_buffer()` in two places:

https://github.com/WordPress/performance/blob/3523e98b62ce2627502f1b75d28c5de1cee48023/modules/images/image-loading-optimization/optimization.php#L159-L167

https://github.com/WordPress/performance/blob/3523e98b62ce2627502f1b75d28c5de1cee48023/modules/images/image-loading-optimization/optimization.php#L317-L319

In the first, it needs to look at all the groups together to see if _any_ of them need URL metrics. This is because at this point we don't know the user's viewport width. So it doesn't make sense here to look up a specific group since there is no specific viewport yet known.

In the second instance, we export the data to `detect.js` so that the client logic (which _does_ know the viewport width) can decide whether or not to send the URL metrics to the server.

So in both of these cases, we need the full data returned by `get_needed_minimum_viewport_widths` and this is re-used in the REST API endpoint via this new `is_group_filled` method.

<!--
In order to get the "breakpoint group ID" (minimum viewport width) you still need to iterate over the minimum viewport widths to figure out which range the given viewport width fits into (now handled by `is_group_filled`). And then you need to look at the groups to figure out if they are filled to the required sample size and/or if all of the URL metrics in each group are not stale (older than the TTL). This is what the `get_needed_minimum_viewport_widths` method does. It could be refactored as you say to instead to first have a method that determines the minimum viewport width for a given viewport width, and then another method that checks if the given group is filled, but-->

Reply:
> So in both of these cases, we need the full data returned by `get_needed_minimum_viewport_widths` and this is re-used in the REST API endpoint via this new `is_group_filled` method.

With the latest changes, these methods are now called `get_viewport_group_lacking_statuses` and `is_viewport_group_lacking`, respectively.

Reply:
As I've been commenting above, I've been working on refactoring this further. I've introduced a `ILO_URL_Metrics_Group_Status` which contains both the `minimum_viewport_width` and the `is_lacking` data. This object is then passed around which I think makes it more clear and readable.

Reply:
@westonruter 
> So in both of these cases, we need the full data returned by `get_needed_minimum_viewport_widths` and this is re-used in the REST API endpoint via this new `is_group_filled` method.

The full list is really only needed in the `ilo_get_detection_script()` call. For the `$needs_detection` check a more intuitive solution would be to just call a method on the class, see https://github.com/WordPress/performance/pull/1010#discussion_r1511727457.

> With the latest changes, these methods are now called `get_viewport_group_lacking_statuses` and `is_viewport_group_lacking`, respectively.

Why? See https://github.com/WordPress/performance/pull/1010#discussion_r1511718561, a positive term would be more intuitive. Whether it's "full" or "filled" doesn't really matter from my perspective, but either is IMO easier to understand and shorter. Negative method names are more complicated, particularly as they may be used with negation.

Reply:
> As I've been commenting above, I've been working on refactoring this further. I've introduced a `ILO_URL_Metrics_Group_Status` which contains both the `minimum_viewport_width` and the `is_lacking` data. This object is then passed around which I think makes it more clear and readable.

See https://github.com/WordPress/performance/pull/1010#discussion_r1511725755 for basically a response to this.